demos/evp/aesccm.c

```diff
@@ -47,79 +48,157 @@ static const unsigned char ccm_tag[] = {
     0x14, 0xdb, 0x50, 0x6d
 };
 
-void aes_ccm_encrypt(void)
+int aes_ccm_encrypt(void)
 {
+    int ret = 0;
     EVP_CIPHER_CTX *ctx;
+    EVP_CIPHER *cipher;
     int outlen, tmplen;
     unsigned char outbuf[1024];
+
     printf("AES CCM Encrypt:\n");
     printf("Plaintext:\n");
     BIO_dump_fp(stdout, ccm_pt, sizeof(ccm_pt));
-    ctx = EVP_CIPHER_CTX_new();
-    /* Set cipher type and mode */
-    EVP_EncryptInit_ex(ctx, EVP_aes_192_ccm(), NULL, NULL, NULL);
+
+    /* Create a context for the encrypt operation */
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL)
+        goto err;
+
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(NULL, "AES-192-CCM", NULL)) == NULL)
+        goto err;
+
+    /* Initialise encrypt operation with the cipher & mode */
+    if (!EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL))
+        goto err;
+
     /* Set nonce length if default 96 bits is not appropriate */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
-                        NULL);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
```

Comment:
Replace these 2 ctrl calls with params... The params can be fed into 
EVP_EncryptInit_ex2

Reply:
Yep, I've tried that. I'm getting a failure when I do this which I think is due to the order in which the arguments & params are processed. Currently investigating this.

Reply:
Try using set_params

Reply:
The params are set after the init code.  This means the IV length is set after the IV has been set.
The code needs to process the IV length before the init code and then set the remaining params.

Reply:
What you talking about Pauli?. It does an initial init with no iv
and then does the controls..

that should be the same as doing init_ex2(no iv, params);

Reply:
It will be the same as that `init_ex2()` call but there is a problem.  The `ccm_init()` function has a sequencing problem:

https://github.com/openssl/openssl/blob/9ac653d81a857a5452f9f25278a24e1dfb226905/providers/implementations/ciphers/ciphercommon_ccm.c#L222-L251

It's setting the params on line 249 after using the IV on line 238.  The `ccm_get_ivlen()` call will be incorrect until the params are processed.  Either the IV length needs to be set by the init function or the IV length param processed first.  I'd prefer the former.

Reply:
I've backed out my changes to ccm_init and now set the parameters in an order which will not cause problems.